File: Python_Ver/PreprocessRawData.py
import numpy as np
import os
from scipy.io import loadmat

from preprocess import *
from utils.spikeTime2Train import spikeTime2Train
from scipy.io import savemat

# Define paths
DATA_INDEX = 'rat01'
DATA_INDEX_SHORT = '01'
ORIGIN_DATA = f'./Python_Ver/raw_data/{DATA_INDEX}.mat'
TARGET_DATA = f'./Python_Ver/data/{DATA_INDEX}.mat'

# Load or preprocess data
if not os.path.exists(TARGET_DATA):
    M1, mPFC, M1num, mPFCnum, M1channelName, mPFCchannelName, segment, trialNo, actions = spikeTime2Train(ORIGIN_DATA)
else:
    data = loadmat(TARGET_DATA)
    M1, mPFC, M1num, mPFCnum = data['M1'], data['mPFC'], data['M1num'].item(), data['mPFCnum']
    segment, trialNo, actions = data['segment'], data['trialNo'], data['actions'][0]

# Rank neurons based on mutual information
_,_,_,M1order = sortM1neurons(M1num, M1, actions)

# Prepare data for decoder training
his = 60  # 600ms history for M1 ensemble
temp = segment[0,(his + 1):] if segment.ndim == 2 else segment[his + 1:]
y = temp[temp > 0]
y_onehot = np.equal(y[:, None], np.arange(1, 4)).astype(int)  # Convert to one-hot encoding

x_full = np.zeros(((his + 1) * M1num, np.sum(temp > 0)))
for i in range(his + 1):
    M1ensemble_temp = M1[(his + 1) - i: -i if i > 0 else None, :].T
    x_full[i * M1num: (i + 1) * M1num, :] = M1ensemble_temp[:, temp > 0]

# Search for best hidden layer size
searchHiddenSize(x_full.T, y_onehot)

# Select optimal hidden size & top-N neurons for decoding
hiddenLayerSize = 2 ** 3
searchM1number(x_full, y_onehot, M1num, M1order, his, hiddenLayerSize)  

# TODO: automatically save the best decoder for RLPP training

# Save preprocessed data
M1num_pre = 2
modelName = f'decodingModel_{DATA_INDEX_SHORT}'

savemat(TARGET_DATA, {
    "DataIndex": DATA_INDEX,
    "his": his,
    "M1": M1,
    "M1num": M1num,
    "M1order": M1order,
    "M1num_pre": M1num_pre,
    "mPFC": mPFC,
    "mPFCnum": mPFCnum,
    "segment": segment,
    "trialNo": trialNo,
    "modelName": modelName,
    "M1channelName": M1channelName,
    "mPFCchannelName": mPFCchannelName,
    "actions": actions
    })

File: PreprocessRawData.m
clear; clc; close all;
addpath utils/ preprocess/

% CHANGE this for six rats
DataIndex = 'rat01';
DataIndexShort = '01';

OriginData = ['raw_data/', DataIndex, '.mat'];
TargetData = ['data/', DataIndex, '.mat'];

%% Load data
if ~exist(TargetData, "file")
    [M1,mPFC,M1num,mPFCnum,M1channelName,mPFCchannelName,segment,trialNo,actions] = spikeTime2Train(OriginData);
else
    load(TargetData, "M1","mPFC","M1num","mPFCnum","segment","trialNo","actions");
end

%% Rank the neurons in decsending order based on the mutual information between neural activity and behavior
sortM1neurons;

%% prepare for decoder training
his = 60; % 600ms history for M1 ensemble
temp=segment((his+1):end);
y=temp(temp>0);
y_onehot=(y'==1:3)'; % only 3 types of behavior

x_full = zeros((his+1)*M1num,sum(temp>0));
for i=0:his
    M1ensemble_temp = M1(((his+1)-i):(end-i),:)';
    x_full(i*M1num+1:(i+1)*M1num,:) = M1ensemble_temp(:,temp>0);
end

%% search the best number of hidden units for the decoder
searchHiddenSize

%% select the proper hidden size & search for the top-N M1 neurons as decoder input
hiddenLayerSize = 2^3; % determined based on the results of last section for each rat
searchM1number

%% select the number of M1 neurons for transregional spike prediction & save the decoding model
M1num_pre = 2; % determined based on the results of last section for each rat

[~,bestIndex]=min(testPercentErrors(M1num_pre,:));
  disp(['Decoding accuracy: ', num2str(1-percentErrors(M1num_pre,bestIndex))]);
% save the best model
modelName = ['decodingModel_',DataIndexShort];
genFunction(netVec{M1num_pre,bestIndex},['decoding/',modelName]);

%% save preprocessed data for transregional spike prediction model traning
save(TargetData, ...
    "DataIndex","his", ...
    "M1","M1num","M1order","M1num_pre", ...
    "mPFC","mPFCnum", ...
    "segment","trialNo", ...
    "modelName");
